cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_cw.cpp

```diff
@@ -0,0 +1,253 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_cw.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+inline auto max_wg(const sycl::queue& q) {
+    return q.get_device().template get_info<sycl::info::device::max_work_group_size>();
+}
+
+template <class Float, class BinaryOp, class UnaryOp>
+class kernel_reduction_rm_cw_inplace {
+    typedef const Float* inp_t;
+    typedef Float* out_t;
+
+public:
+    kernel_reduction_rm_cw_inplace(inp_t const input_,
+                                   out_t const output_,
+                                   const std::int32_t width_,
+                                   const std::int64_t height_,
+                                   const std::int32_t lstride_,
+                                   const BinaryOp binary_,
+                                   const UnaryOp unary_)
+            : input{ input_ },
+              output{ output_ },
+              unary{ unary_ },
+              binary{ binary_ },
+              width{ width_ },
+              height{ height_ },
+              lstride{ lstride_ } {}
+
+    void operator()(sycl::nd_item<2> it) const {
+        using sycl::ONEAPI::reduce;
+        // Common for whole WG
+        const auto col_idx = it.get_global_id(0);
```

Comment:
The same comment about col/row

Reply:
Fixed. It is a kernel itself. It works with low-level abstractions not with ndarray